### PR TITLE
Add CONTRIBUTING section on build

### PR DESCRIPTION
**Pull request by thockin**
@kubernetes/contributor-experience

---
This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/36590)

### CODE REVIEW

**Comment by chrislovecnm** [comment]
YAY ... assigned to a PR :)

File: CONTRIBUTING.md [hunk 1]
```diff
@@ -82,6 +82,27 @@ some bugs or smaller features.  We have a [feature development
 process](https://github.com/kubernetes/features/blob/master/README.md), but
 navigating the Kubernetes system as a newcomer can be very challenging.
 
+### Building the project
+
+There are a few things you need to build and test Kubernetes:
```
Comment (chrislovecnm) [inline CONTRIBUTING.md]: > Build and test Kubernetes

This project?

**Comment by grodrigues3** [comment]
lgtm

File: CONTRIBUTING.md [hunk 2]
```diff
@@ -82,6 +82,27 @@ some bugs or smaller features.  We have a [feature development
...
+1. `make` - the human interface to the Kubernetes build is `make`, so you must
+   have this tool installed on your machine.  We try not to use too many crazy
+   features of `Makefile`s and other tools, so most commonly available versions
+   should work.
+1. `docker` - some parts of the build/test system depend on `docker`.  You
+   need a relatively recent version of Docker installed, and available to you.
+1. `go` - Kubernetes is written in Go (aka golang), so you need a relatively
+   recent version of the Go toolchain installed.
```
Comment (chrislovecnm) [inline CONTRIBUTING.md]: Which platforms does this work on?  Windows?  Also should we link a n00b doc that has links to install instructions for each.  I always approach this as a really really dumb perspective :)

File: CONTRIBUTING.md [hunk 3]
```diff
@@ -82,6 +82,27 @@ some bugs or smaller features.  We have a [feature development
...
+
+To build Kubernetes, simply type `make`.  This should figure out what it needs
+to do and not need any input from you.  If you want to just build a subset of
+code, you can pass the `WHAT` variable to `make`: e.g. `make
+WHAT="pkg/kubelet"`.
```
Comment (dims) [inline CONTRIBUTING.md]: "cmd/kubectl" may be a better example as it generates a binary folks can use!

Comment (chrislovecnm) [inline CONTRIBUTING.md]:
```console
$ make
```

File: CONTRIBUTING.md [hunk 4]
```diff
@@ -82,6 +82,27 @@ some bugs or smaller features.  We have a [feature development
...
+1. `make` - the human interface to the Kubernetes build is `make`, so you must
+   have this tool installed on your machine.  We try not to use too many crazy
+   features of `Makefile`s and other tools, so most commonly available versions
+   should work.
+1. `docker` - some parts of the build/test system depend on `docker`.  You
+   need a relatively recent version of Docker installed, and available to you.
+1. `go` - Kubernetes is written in Go (aka golang), so you need a relatively
+   recent version of the Go toolchain installed.
+
+To build Kubernetes, simply type `make`.  This should figure out what it needs
```
Comment (chrislovecnm) [inline CONTRIBUTING.md]: Type on a command line...

File: CONTRIBUTING.md [hunk 5]
```diff
@@ -82,6 +82,27 @@ some bugs or smaller features.  We have a [feature development
...
+to do and not need any input from you.  If you want to just build a subset of
+code, you can pass the `WHAT` variable to `make`: e.g. `make
+WHAT="pkg/kubelet"`.
+
+To run basic tests, simply type `make test`.  This will run all of the unit
+tests in the project.  If you want to just test a subset of the project, you
+can pass the `WHAT` variable to `make`: e.g. `make test WHAT=pkg/kubelet`.
```
Comment (chrislovecnm) [inline CONTRIBUTING.md]: > variable

environment variable :)

**Review by chrislovecnm — CHANGES_REQUESTED** [review]
A couple of tweaks to talk about.  Excited that we are improving this!!!

File: CONTRIBUTING.md [hunk 6]
```diff
@@ -82,6 +82,27 @@ some bugs or smaller features.  We have a [feature development
...
 
+### Building the project
+
+There are a few things you need to build and test Kubernetes:
+1. `make` - the human interface to the Kubernetes build is `make`, so you must
+   have this tool installed on your machine.  We try not to use too many crazy
+   features of `Makefile`s and other tools, so most commonly available versions
+   should work.
+1. `docker` - some parts of the build/test system depend on `docker`.  You
+   need a relatively recent version of Docker installed, and available to you.
```
Comment (philips) [inline CONTRIBUTING.md]: this needs a link to our recommendation. I would suggest minikube onboarding and docker-env.

File: CONTRIBUTING.md (hunk 4 above)
Comment (philips) [inline CONTRIBUTING.md]: You will need to git clone it first too :)

**Review by philips — COMMENTED** [review]
Improvement

**Comment by spxtr** [comment]
It might be worth adding a line saying that we are investigating bazel with a link to the doc about that.

**Comment by chrislovecnm** [comment]
@spxtr I don't want to talk about how the build runs, I would recommend talking about how to run the build :)  Bazel is awesome, but can we document that separately.

File: CONTRIBUTING.md (hunk 1 above)
Comment (spxtr) [inline CONTRIBUTING.md]: You need a newline after this or the list won't render properly.

**Comment by spxtr** [comment]
@chrislovecnm that's reasonable.

**Comment by pigmej** [comment]
I have some suggestion can we somehow recommend using https://github.com/Mirantis/k8s-devbox and/or https://github.com/Mirantis/kubeadm-dind-cluster as a *fast* way to test your stuff? devbox is supposed to create reproducible development environment for working on Kubernetes project and there are some plans to add kubeadm to it which will allow to play with remote clusters too. I'm open to suggestions there. (cc @ivan4th author of these).

File: CONTRIBUTING.md (hunk 6 above)
Comment (ivan4th) [inline CONTRIBUTING.md]: [According to Dan Lorenc](https://groups.google.com/d/msg/kubernetes-sig-cluster-lifecycle/XFODRr_TXgE/hmJ_5v5LCgAJ), minikube is targeting k8s end users. Using it for development of Kubernetes may be [not very convenient](https://github.com/kubernetes/minikube/blob/master/UPDATE_KUBERNETES.md). Or did I miss something here? Also, minikube by itself is probably overkill if the developer just wants to have docker.

**Comment by ivan4th** [comment]
@jnowak: to be more precise, k8s-devbox tries to help getting k8s development environment up&running quickly and also tries to help new developers getting acquinted with k8s development workflow. As of now it uses [kubernetes-dind-cluster](https://github.com/sttts/kubernetes-dind-cluster) as a fast method of *multinode* test cluster startup, but it will support [kubeadm-dind-cluster](https://github.com/Mirantis/kubeadm-dind-cluster) soon. The latter uses kubeadm to create docker-in-docker-based cluster and supports workflows based on remote docker engine (copy binaries directly from the build data container & run the test cluster on the same remote docker where they were built). Both of these projects support running e2e tests locally, including ones requiring multiple nodes (unlike local-up-cluster). They're also much faster than VM-based clusters.

File: CONTRIBUTING.md (hunk 6 above)
Comment (pigmej) [inline CONTRIBUTING.md]: +1 for Ivan, minikube is *not* for developers, for developers look at https://github.com/kubernetes/kubernetes/pull/36590#issuecomment-259802736 or https://github.com/kubernetes/kubernetes/pull/36590#issuecomment-259808686

File: CONTRIBUTING.md (hunk 3 above)
Comment (thockin) [inline CONTRIBUTING.md]: Done

File: CONTRIBUTING.md (hunk 1 above)
Comment (thockin) [inline CONTRIBUTING.md]: Done

File: CONTRIBUTING.md (hunk 4 above)
Comment (thockin) [inline CONTRIBUTING.md]: ok, adding some words about downloads.

File: CONTRIBUTING.md (hunk 3 above)
Comment (thockin) [inline CONTRIBUTING.md]: disagree - this is a commandline, not a console log.

File: CONTRIBUTING.md (hunk 5 above)
Comment (thockin) [inline CONTRIBUTING.md]: This is not an env var, it is a make var.  I think env vars also work, but that is not what is being shown here

**Comment by thockin** [comment]
All, this doc is NOT an encyclopedic listing of everythign you need to know about how to use kube.  It is an intro doc.  Feel free to send followups that link to more details.

File: CONTRIBUTING.md (hunk 5 above)
Comment (chrislovecnm) [inline CONTRIBUTING.md]: Touche

File: CONTRIBUTING.md (hunk 3 above)
Comment (chrislovecnm) [inline CONTRIBUTING.md]: Sounds good

**Comment by chrislovecnm** [comment]
@thockin do we need an issue for an encyclopedic version of this doc?

**Comment by thockin** [comment]
Probably.  I am on mobile now, can do later

**Comment by bgrant0607** [comment]
@thockin It shouldn't block this PR, but I'd like this doc to be very brief, with links to more information, which won't live in the kubernetes repo.

File: CONTRIBUTING.md [hunk 7]
```diff
@@ -82,6 +82,46 @@ some bugs or smaller features.  We have a [feature development
...
+1. `go` - Kubernetes is written in Go (aka golang), so you need a relatively
+   recent version of the [Go toolchain](https://golang.org/dl/) installed.
+   While Linux is the primary platform for Kubernetes, it should compile on a
+   Mac, too.  Windows is in progress.
+
+To build Kubernetes, simply type `make`.  This should figure out what it needs
```
Comment (bgrant0607) [inline CONTRIBUTING.md]: #36655 claims that this doesn't work.

**Comment by bgrant0607** [comment]
cc @sebgoa

**Comment by bgrant0607** [comment]
Let's merge and iterate.
LGTM.

File: CONTRIBUTING.md (hunk 7 above)
Comment (sebgoa) [inline CONTRIBUTING.md]: it does not work, if you don't have a go environment setup and use the documented `build-tools.sh/make`

File: CONTRIBUTING.md (hunk 6 above)
Comment (philips) [inline CONTRIBUTING.md]: Sure "but pretty much any recent version" is not sufficient and I would really just like us to punt to somewhere where we know a known good version is being maintained. For cross-platform availability for workstations the closest thing we got is minikube.

@ivan4th @pigmej Also, it isn't for k8s developer but it is a known good version of the Docker engine that the k8s community tests and packages as VMs.

**Review by philips — COMMENTED** [review]
Overall I am not going to fight this Docker version thing but we are honestly too close to the tech here to think "just install some known good version of Docker" is good enough as a developer onramp.

**Comment by thockin** [comment]
If I knew where to punt, I would

**Comment by philips** [comment]
@thockin I am suggesting we punt to minikube and force minikube to have a good docker install story on all platforms.

**Pull request merged**